tools/codegen/dest/native_functions.py

```diff
@@ -1,96 +1,59 @@
-from typing import List, Union, Set, Any
+from typing import List, Union, Optional
 
-from tools.codegen.context import with_native_function
-from tools.codegen.utils import concatMap
-from tools.codegen.model import (NativeFunction, NativeFunctionsGroup,
-                                 ExternalBackendFunction, ExternalBackendFunctionsGroup,
-                                 is_structured_dispatch_key)
-from tools.codegen.api.types import DispatcherSignature, NativeSignature
+from tools.codegen.context import with_native_function_and_index
+from tools.codegen.utils import mapMaybe
+from tools.codegen.model import NativeFunction, NativeFunctionsGroup, BackendIndex
+from tools.codegen.api.types import kernel_signature
 import tools.codegen.api.meta as meta
-import tools.codegen.api.native as native
 import tools.codegen.api.structured as structured
 
-@with_native_function
-def gen_unstructured(f: NativeFunction) -> List[str]:
-    ns = list(f.dispatch.values())
-    native_sig = NativeSignature(f.func)
-
-    rs = []
-    # Sometimes a function name shows up multiple times; only generate
-    # it once!
-    seen = set()
-    for n in ns:
-        if n in seen:
-            continue
-        if "legacy::" in n:
-            continue
-        seen.add(n)
-        rs.append(f"TORCH_API {native_sig.decl(name=n)};")
-
-    return rs
-
-@with_native_function
-def gen_unstructured_external(f: ExternalBackendFunction) -> List[str]:
-    # XLA appears to have used the dispatcher convention to write their kernel signatures,
-    # probably because they based their signatures off of our RegistrationDeclarations.h
-    dispatcher_sig = DispatcherSignature.from_schema(f.native_function.func)
-    if f.metadata is not None:
-        # Only generate declarations for operators that xla has defined in the yaml
-        return [f"static {dispatcher_sig.decl()};"]
+@with_native_function_and_index
+def gen_unstructured(f: NativeFunction, backend_index: BackendIndex) -> Optional[str]:
+    sig = kernel_signature(f, backend_index)
+    name = backend_index.kernel(f)
+    is_external = backend_index.is_external(f)
+    # sigh... I should be able to better enforce that these are all or nothing.
+    if sig is None or name is None or is_external is None:
+        return None
+    if "legacy::" in name:
+        return None
     else:
-        return []
+        prefix = 'static' if is_external else 'TORCH_API'
```

Comment:
hum, I wonder why we mark these `TORCH_API` in core. I guess it's because people access these with `native::`...